enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/InternalFlatFileRealmTest.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.shiro.realm.Realm;
+import org.apache.shiro.subject.PrincipalCollection;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.time.Clock;
+import java.util.List;
+
+import org.neo4j.kernel.api.security.exception.InvalidAuthTokenException;
+import org.neo4j.server.security.auth.AuthenticationStrategy;
+import org.neo4j.server.security.auth.BasicPasswordPolicy;
+import org.neo4j.server.security.auth.InMemoryUserRepository;
+import org.neo4j.server.security.auth.PasswordPolicy;
+import org.neo4j.server.security.auth.RateLimitedAuthenticationStrategy;
+import org.neo4j.server.security.auth.UserRepository;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+import static org.neo4j.server.security.auth.SecurityTestUtils.authToken;
+import static org.neo4j.server.security.enterprise.auth.AuthTestUtil.listOf;
+
+public class InternalFlatFileRealmTest
+{
+    private MultiRealmAuthManager authManager;
+    private TestRealm testRealm;
+
+    @Before
+    public void setup() throws Throwable
+    {
+        testRealm = new TestRealm(
+                        new InMemoryUserRepository(),
+                        new InMemoryRoleRepository(),
+                        new BasicPasswordPolicy(),
+                        new RateLimitedAuthenticationStrategy( Clock.systemUTC(), 3 ) );
+
+        List<Realm> realms = listOf( testRealm );
+
+        authManager = new MultiRealmAuthManager( testRealm, realms, new MemoryConstrainedCacheManager() );
+        authManager.init();
+        authManager.start();
+
+        authManager.getUserManager().newUser( "mike", "123", false );
+    }
+
+    @Test
+    public void shouldNotCacheAuthenticationInfo() throws InvalidAuthTokenException
+    {
+        // Given
+        authManager.login( authToken( "mike", "123" ) );
+        assertThat( "Test realm did not receive a call", testRealm.wasAuthenticated(), is( true ) );
+
+        // When
+        authManager.login( authToken( "mike", "123" ) );
+
+        // Then
+        assertThat( "Test realm did not receive a call", testRealm.wasAuthenticated(), is( true ) );
+    }
+
+    @Test
+    public void shouldNotCacheAuthorizationInfo() throws InvalidAuthTokenException
+    {
+        // Given
+        EnterpriseAuthSubject mike = authManager.login( authToken( "mike", "123" ) );
+        mike.allowsReads();
+        assertThat( "Test realm did not receive a call", testRealm.wasAuthorized(), is( true ) );
+
+        // When
+        mike.allowsWrites();
+
+        // Then
+        assertThat( "Test realm did not receive a call", testRealm.wasAuthorized(), is( true ) );
+    }
+
+    private class TestRealm extends InternalFlatFileRealm
+    {
+        private boolean authenticated = false;
+        private boolean authorized = false;
+
+        boolean wasAuthenticated()
+        {
+            boolean t = authenticated;
+            authenticated = false;
```

Comment:
I found these tests slightly hard to follow because of these side-effects (here and wasAuthorized()). Maybe we could make it more explicit?